### PR TITLE
docs: Add SKD014/SKD015 hyperparameter search pitfall example

### PR DESCRIPTION
**Pull request by moujanrastgoo**
#### Change description
Related to #2622 and #3134  
- Adds new sphinx-gallery subsection **Pitfalls and solutions**
- Registers subsection in `sphinx/conf.py`
- Adds SKD014 and SKD015 check example and links from automated checks user guide

#### Contribution checklist

- [ ] Unit tests were added or updated (if necessary)
- [x] Documentation was added or updated (if necessary)
- [x] The code passes our style conventions (you can check this by running pre-commit on your code
with `pre-commit run --all-files`)
- [ ] All the tests pass (please test locally before pushing)
- [ ] The documentation builds and renders properly (if it does, our bot will add a comment linking
to a preview of the documentation to review it visually)
- [x] All the commits in the PR are signed (more information
[here](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits))
- [x] The pull request title respects the Conventional Commits convention (more information
[here](https://docs.skore.probabl.ai/stable/contributing.html#pull-requests))

#### AI usage disclosure

AI tools were involved for:
- [x] Code generation (e.g., when writing an implementation or fixing a bug)
- [ ] Test/benchmark generation
- [x] Documentation (including examples)
- [x] Research and understanding

### CODE REVIEW

**Comment by moujanrastgoo** [comment]
Working on the last section of the example with randomized search.

File: examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+
+SKD014 & SKD015: Hyperparameter search pitfalls
+===============================================
+
+This example walks through mitigations when checks
+:ref:`SKD014 <skd014-hyperparams-at-search-edge>` and
+:ref:`SKD015 <skd015-hyperparameters-worth-tuning>` fire on a fitted search
+object. SKD014 is an issue: numeric ``best_params_`` sit on the minimum or
+maximum value tried, so the true optimum may lie outside the searched range.
+SKD015 is a tip: high-impact axes are missing from the search space, which
```
Comment (glemaitre) [inline examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py]: high-impact axes is not easy to know what it is. I think the following is better:


```suggestion
SKD015 is a tip: an hyperparameter is missing from the search space, which
```

File: examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+is incomplete rather than necessarily wrong.
+
+These findings often appear together on the same
+:class:`~sklearn.model_selection.BaseSearchCV` report. The user guide
+recommends addressing them jointly when both fire.
```
Comment (glemaitre) [inline examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py]:
```suggestion
Usually, those tests are related to the search CV object and we advocate to address them
jointly when both fire.
```

File: examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+- extend ``param_grid`` or ``param_distributions`` beyond the flagged bounds,
+- for :class:`~sklearn.model_selection.RandomizedSearchCV`, increase ``n_iter``
+  and sample from a wider range,
+- if SKD015 also fires, widen the search on every recommended axis.
+
+**SKD015: hyperparameters worth tuning** (tip)
+
+- add the suggested parameters to ``param_grid`` or ``param_distributions``.
+
+We tune a :class:`~sklearn.ensemble.HistGradientBoostingClassifier` inside
```
Comment (glemaitre) [inline examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py]:
```suggestion
In this example, we tune a :class:`~sklearn.ensemble.HistGradientBoostingClassifier` inside
```

File: examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+that clears both.
+"""
+
+# %%
+# Load the employee salaries dataset
+# ==================================
+#
+# We predict whether salary exceeds the median. Mixed HR features suit
+# :func:`~skrub.tabular_pipeline`. A 3,000-row stratified subsample keeps the
+# gallery grids short while preserving class balance.
```
Comment (glemaitre) [inline examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py:47]: We should be mentioning that we transform the regression problem into a classification one where we want to predict whether or not an employee earn more than the median salary of employees.

File: examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+)
+
+X, _, y, _ = train_test_split(
+    X_full,
+    y_full,
+    train_size=3_000,
+    stratify=y_full,
+    random_state=42,
+)
+y = pd.Series(y, name="high_earner")
```
Comment (glemaitre) [inline examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py]: We can simplify the code here:

```suggestion
dataset = fetch_employee_salaries()
X_full = dataset.X
y_full = (dataset.y > dataset.y.median()).astype(int).rename("high_earner")

X, _, y, _ = train_test_split(
    X_full,
    y_full,
    train_size=3_000,
    stratify=y_full,
    random_state=42,
)
```

File: examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+# %%
+# Shared search setup
+# ===================
+#
+# The base object is ``tabular_pipeline`` around HGB. Outer hold-out and inner
+# ``cv=3`` folds are stratified (no time ordering required for these checks).
+
+from sklearn.ensemble import HistGradientBoostingClassifier
+from sklearn.model_selection import GridSearchCV
+from skore import TrainTestSplit
```
Comment (glemaitre) [inline examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py]: Delay this import to when it is called.

File: examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+from skrub import tabular_pipeline
+
+splitter = TrainTestSplit(test_size=0.2, random_state=42, stratify=y)
+cv = 3
```
Comment (glemaitre) [inline examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py]: Here, it is as well confusing that we have both parameter `splitter` and `cv`. We should only use `splitter`. But it means that we need to declare those parameters just before to import them.

File: examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+base_pipeline = tabular_pipeline(
+    HistGradientBoostingClassifier(max_iter=200, random_state=42)
+)
```
Comment (glemaitre) [inline examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py:97]:
```suggestion
base_pipeline = tabular_pipeline(
    HistGradientBoostingClassifier(max_iter=200, random_state=42)
)
base_pipeline
```

File: examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py (hunk 6 above)
Comment (glemaitre) [reply examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py]: The same for the grid search

File: examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+# ``max_iter`` is a budget parameter, not a complexity knob in the SKD015 table.
+# A single-value grid still runs a search, but SKD014 skips axes with fewer than
+# two distinct tried values, so this beat isolates the SKD015 tip.
+
+from skore import evaluate
+
+max_iter_only_search = GridSearchCV(
+    base_pipeline,
+    param_grid={
+        "histgradientboostingclassifier__max_iter": [200],
```
Comment (glemaitre) [inline examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py]: I think that it would make sense to have at least 2 values in the grid.
Otherwise, we don't necessary know why we need a grid search.

File: examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+    },
+    cv=cv,
+    n_jobs=4,
```
Comment (glemaitre) [inline examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py:131]: Here, we have the issue that we are optimizing the accuracy because it the default skore. I think that we should always use the log-loss and mean squared error.

File: examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+    refit=True,
+)
+
+report_axes = evaluate(
```
Comment (glemaitre) [inline examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py]: I think that we should call it `report`. `_axes` is a weird name.

File: examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+    max_iter_only_search,
+    X=X,
+    y=y,
+    pos_label=1,
+    splitter=splitter,
+)
+
+# %%
+# SKD015 should tip that learning rate, depth, and leaf size were not searched.
```
Comment (glemaitre) [inline examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py]: We should explicitely mentioned that it is inside the ("Tips" tab)

File: examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+# ``best_params_`` only contains ``max_iter`` - that incompleteness is the point.
+# A search that only tweaks training budget ignores the axes that usually move
```
Comment (glemaitre) [inline examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py]: I think that we need to find an alternative word to `axes`. So potentially just refer to hyperparameter.

File: examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+report_edge.checks.summarize(fast_mode=True)
+
+# %%
+report_edge.estimator_.best_params_
+
+# %%
+report_edge.metrics.summarize(data_source="both").frame()
+
+# %%
+# SKD014 & SKD015: widen bounds and add recommended axes
```
Comment (glemaitre) [inline examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py]: For here and after, let's rename axes.

File: examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+
+from skore import compare
+
+compare(
+    {
+        "max_iter_only": report_axes,
+        "two_point_edge_grid": report_edge,
+        "padded_recommended_axes": report_full,
+    }
+).metrics.summarize(data_source="both").frame()
```
Comment (glemaitre) [inline examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py]: Potentially, we would to translate the matrix and limit the score to roc auc and log loss.

**Review by glemaitre — COMMENTED** [review]
Here is a round of comments

File: examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+TableReport(X)
+
+# %%
+TableReport(y)
+
+# %%
+# Shared search setup
+# ===================
+#
+# Let us wrap HGB in ``tabular_pipeline``. Early stopping lets a wide
```
Comment (glemaitre) [inline examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py]:
```suggestion
# Let us wrap HGB in :func:`~skore.tabular_pipeline`. Early stopping lets a wide
```

File: examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py [hunk 17]
```diff
@@ -0,0 +1,273 @@
...
+# ``max_iter`` grid pick an interior budget in the first beat below. The outer
+# hold-out uses :class:`~skore.TrainTestSplit` when we call
+# :func:`~skore.evaluate`; each ``GridSearchCV`` below sets its own inner
```
Comment (glemaitre) [inline examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py]:
```suggestion
# :func:`~skore.evaluate`; each :class:`~sklearn.model_selection.GridSearchCV` below sets its own inner
```

File: examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py [hunk 18]
```diff
@@ -0,0 +1,273 @@
...
+
+# %%
+# Compare search strategies
+# =========================
+#
+# Hold-out log-loss and ROC AUC for the incomplete grid, the two-point edge
+# grid, and the joint fix. Clearing the findings means the *search design*
+# improved; still judge models on validation metrics and cost, not check status
+# alone. ``transpose()`` flips the table so each search strategy is a row and
+# each metric is a column.
```
Comment (glemaitre) [inline examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py]:
```suggestion
# alone.
```

File: examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py [hunk 19]
```diff
@@ -0,0 +1,273 @@
...
+
+# %%
+# Conclusion
+# ==========
+#
+# SKD015 tips about incomplete search spaces; SKD014 issues when optima stick to
+# the boundary of the box you tried. In this walkthrough, a ``max_iter``-only
+# search missed key hyperparameters, a two-point grid forced edge hits, and one
+# padded complete grid addressed both. Expand the search before deploying
+# ``best_params_``: green checks are about search hygiene, not a guarantee of
```
Comment (glemaitre) [inline examples/pitfalls_and_solutions/plot_skd014_skd015_hyperparameter_search.py]: are the check really "green" :)

**Review by glemaitre — DISMISSED** [review]
Otherwise LGTM.

**Review by glemaitre — APPROVED** [review]
LGTM. Thanks @moujanrastgoo merging.
